scripts/validate_rst_title_capitalization.py

```diff
@@ -0,0 +1,211 @@
+    "Arrow",
+    "Parquet",
+    "MultiIndex",
+    "NumFOCUS",
+    "sklearn",
+    "Docker",
+}
+
+CAP_EXCEPTIONS_DICT = {word.lower(): word for word in CAPITALIZATION_EXCEPTIONS}
+
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
+
+
+def correct_title_capitalization(title: str) -> str:
+    """
+    Algorithm to create the correct capitalization for a given title.
+
+    Parameters
+    ----------
+    title : str
+        Heading string to correct.
+
+    Returns
+    -------
+    correct_title : str
+        Correctly capitalized heading.
+
+    """
+
+    # Function to strip all non-word characters from the beginning of the title to the
+    # first word character.
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
+
+    # Function to remove a URL from the title. We do this because words in a URL must
+    # stay lowercase, even if they are a capitalization exception.
+    removed_https_title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", correct_title)
+
+    # Function to split a title into a list using non-word character delimiters.
+    word_list = re.split(r"\W", removed_https_title)
+
+    for word in word_list:
+        if word.lower() in CAP_EXCEPTIONS_DICT:
+            correct_title = re.sub(
+                rf"\b{word}\b", CAP_EXCEPTIONS_DICT[word.lower()], correct_title
+            )
+
+    return correct_title
+
+
+def find_titles(rst_file: str) -> Generator[Tuple[str, int], None, None]:
+    """
+    Algorithm to identify particular text that should be considered headings in an
+    RST file.
+
+    See <https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html> for details
+    on what constitutes a string as a heading in RST.
+
+    Parameters
+    ----------
+    rst_file : str
+        RST file to scan through for headings.
+
+    Yields
+    -------
+    title : str
+        A heading found in the rst file.
+
+    line_number : int
+        The corresponding line number of the heading.
+
+    """
+
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
+
+    symbols = ("*", "=", "-", "^", "~", "#", '"')
+
+    table = str.maketrans("", "", "*`_")
+
+    for i, line in enumerate(lines):
+        line_chars = set(line)
+        if (
+            len(line_chars) == 1
+            and line_chars.pop() in symbols
+            and len(line) == len(lines[i - 1])
+        ):
+            yield lines[i - 1].translate(table), i
+
+
+def find_rst_files(source_paths: List[str]) -> Generator[str, None, None]:
+    """
+    Given the command line arguments of directory paths, this method
+    yields the strings of the .rst file directories that these paths contain.
+
+    Parameters
+    ----------
+    source_paths : str
+        List of directories to validate, provided through command line arguments.
+
+    Yields
+    -------
+    directory_address : str
+        Directory address of a .rst files found in command line argument directories.
+
```

Comment:
Better remove this blank line (we're not validating this docstring, but this blank line would make it fail if we do.